src/ape/managers/_contractscache.py

```diff
@@ -606,6 +591,38 @@ def get(
 
         return contract_type
 
+    def _get_proxy_contract_type(
+        self,
+        address: AddressType,
+        proxy_info: ProxyInfoAPI,
+        fetch_from_explorer: bool = True,
+        default: Optional[ContractType] = None,
+    ) -> Optional[ContractType]:
+        implementation_contract_type = self.get(proxy_info.target, default=default)
+        proxy_contract_type = self.get(
+            address,
+            fetch_from_explorer=fetch_from_explorer,
+            proxy_info=proxy_info,
+            detect_proxy=False,
+        )
+        if proxy_contract_type is not None and implementation_contract_type is not None:
+            combined_contract = _get_combined_contract_type(
+                proxy_contract_type, proxy_info, implementation_contract_type
+            )
+            self.contract_types[address] = combined_contract
+            return combined_contract
+
+        elif implementation_contract_type is not None:
+            contract_type_to_cache = implementation_contract_type
+            self.contract_types[address] = implementation_contract_type
+            return contract_type_to_cache
+
+        elif proxy_contract_type is not None:
+            self.contract_types[address] = proxy_contract_type
+            return proxy_contract_type
+
+        return None
```

Comment:
Maybe consider a WARNING log here, something like "couldn't determine any contract_type for proxy {address}"